### PR TITLE
Decompose Adjustment into smaller steps and remove the method map. 

### LINKED ISSUE #3548: port the "autoref-arg" code used in overloaded operators to use the ptr adjustment table
We have this nice table in the tcx called `adjustments` that's supposed to track when an expression is being auto-ref'd or auto-deref'd.  It would make borrowck, trans, and other pieces of code that are concerned with this "just work".  But it's not being used for overloaded operators right now.

### PR DESCRIPTION
**Pull request by eddyb**
The method map held method callee information for:
* actual method calls (`x.f(...)`)
* overloaded unary, binary, indexing and call operators
* *every overloaded deref adjustment* (many can exist for each expression)

That last one was a historical ~~accident~~ hack, and part of the motivation for this PR, along with:
* a desire to compose adjustments more freely
* containing the autoderef logic better to avoid mutation within an inference snapshot
* not creating `TyFnDef` types which are incompatible with the original one
  * i.e. we used to take a`TyFnDef`'s `for<'a> &'a T -> &'a U` signature and instantiate `'a` using a region inference variable, *then* package the resulting `&'b T -> &'b U` signature in another `TyFnDef`, while keeping *the same* `DefId` and `Substs`
* to fix #3548 by explicitly writing autorefs for the RHS of comparison operators

Individual commits tell their own story, of "atomic" changes avoiding breaking semantics.

Future work based on this PR could include:
* removing the signature from `TyFnDef`, now that it's always "canonical"
  * some questions of variance remain, as subtyping *still* treats the signature differently
* moving part of the typeck logic for methods, autoderef and coercion into `rustc::traits`
* allowing LUB coercions (joining multiple expressions) to "stack up" many adjustments
* transitive coercions (e.g. reify or unsize after multiple steps of autoderef)

r? @nikomatsakis

### CODE REVIEW

**Comment by eddyb** [comment]
Started crater run.

**Comment by eddyb** [comment]
The current Travis failures are in UI tests due to error output changed (intentionally), e.g.:
```diff
 error[E0277]: the trait bound `{integer}: std::ops::Mul<()>` is not satisfied
-  --> $DIR/binops.rs:14:5
+  --> $DIR/binops.rs:14:7
    |
 14 |     3 * ();
-   |     ^^^^^^ no implementation for `{integer} * ()`
+   |       ^ no implementation for `{integer} * ()`
    |
    = help: the trait `std::ops::Mul<()>` is not implemented for `{integer}`
```

cc @rust-lang/docs @nikomatsakis @jonathandturner Is this is an improvement?

**Comment by GuillaumeGomez** [comment]
@eddyb: I don't think it is. It should still underline the whole expression.

**Comment by Mark-Simulacrum** [comment]
I personally consider it an improvement, as it points at the thing that isn't implemented. The operands aren't really part of the problem; making the span as small as possible seems better.

**Comment by eddyb** [comment]
[Crater report](https://gist.github.com/eddyb/626d1fd2b229c28ea96184b3950f12de) is clean (modulo the usual network errors and lint plugins).

**Comment by QuietMisdreavus** [comment]
I'm with @Mark-Simulacrum here. The error text still says which specific kind of multiplication isn't implemented, and pointing directly at the operator can save some confusion if it were part of a larger/more complicated expression.

**Comment by jonathandturner** [comment]
@eddyb - yes, pointing at the offending operator looks like an improvement to me.  As @QuietMisdreavus points out, this becomes more important with larger expressions, but even small ones seem to be a little clearer

**Comment by eddyb** [comment]
Thanks everyone! I pushed an update to the UI tests.

File: src/librustc/ty/adjustment.rs [hunk 1]
```diff
@@ -13,10 +13,83 @@ use hir::def_id::DefId;
 use ty::{self, Ty, TyCtxt, TypeAndMut};
 use ty::subst::Substs;
 
+
+/// Represents coercing a value to a different type of value.
+///
+/// We transform values by following the following steps in order:
+/// 1. Apply a step of `Adjust` (see its variants for details).
```
Comment (nikomatsakis) [inline src/librustc/ty/adjustment.rs]: this should say "apply the base transformation described by `kind`", right?

File: src/librustc/ty/context.rs [hunk 2]
```diff
@@ -341,15 +341,20 @@ impl<'tcx> TypeckTables<'tcx> {
...
     /// entry recorded for that expression.
     pub fn expr_ty_adjusted(&self, expr: &hir::Expr) -> Ty<'tcx> {
-        self.adjustments.get(&expr.id)
+        self.expr_adjustments(expr).last()
             .map_or_else(|| self.expr_ty(expr), |adj| adj.target)
     }
 
     pub fn expr_ty_adjusted_opt(&self, expr: &hir::Expr) -> Option<Ty<'tcx>> {
-        self.adjustments.get(&expr.id)
+        self.expr_adjustments(expr).last()
```
Comment (nikomatsakis) [inline src/librustc/ty/context.rs]: Nit: I'd find this easier to read if each step in the chain (i.e., `.last()`, `.map()`, and `.or_else()` were on their own line). Or at least `.map()` and `.or_else()`.

File: src/libcollections/tests/binary_heap.rs [hunk 3]
```diff
@@ -134,22 +134,22 @@ fn test_push() {
 fn test_push_unique() {
     let mut heap = BinaryHeap::<Box<_>>::from(vec![box 2, box 4, box 9]);
     assert_eq!(heap.len(), 3);
-    assert!(*heap.peek().unwrap() == box 9);
+    assert!(**heap.peek().unwrap() == 9);
```
Comment (nikomatsakis) [inline src/libcollections/tests/binary_heap.rs]: what's happening here? some lints firing?

File: src/test/ui/impl-trait/equality.stderr [hunk 4]
```diff
@@ -8,10 +8,10 @@ error[E0308]: mismatched types
               found type `u32`
 
 error[E0277]: the trait bound `u32: std::ops::Add<impl Foo>` is not satisfied
-  --> $DIR/equality.rs:34:9
+  --> $DIR/equality.rs:34:11
    |
 34 |         n + sum_to(n - 1)
-   |         ^^^^^^^^^^^^^^^^^ no implementation for `u32 + impl Foo`
+   |           ^ no implementation for `u32 + impl Foo`
```
Comment (nikomatsakis) [inline src/test/ui/impl-trait/equality.stderr]: nice

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
a few nits, but r=me. Very cool!

**Comment by nikomatsakis** [comment]
Regarding the new spans -- yes, I consider it an improvement. I believe this is what clang does. Part of the reason I think this is better: mutil-line spans are terrible, and this avoids that.

**Comment by nikomatsakis** [comment]
>  some questions of variance remain, as subtyping still treats the signature differently

I was wondering what happened there.

File: src/libcollections/tests/binary_heap.rs (hunk 3 above)
Comment (eddyb) [inline src/libcollections/tests/binary_heap.rs]: Yupp! Comparison operators are now understood by the "unnecessary allocation" lint.

**Comment by eddyb** [comment]
@nikomatsakis I actually rebased those on top of this branch today, but tried to keep this PR separate.

File: src/librustc/ty/adjustment.rs (hunk 1 above)
Comment (eddyb) [inline src/librustc/ty/adjustment.rs]: What do you mean by "base"? But yeah I could use `kind` and then refer to "variants of `Adjust`".

Comment (eddyb) [inline src/librustc/ty/adjustment.rs]: Oh look at the latest version of this comment, that was transient.

**Pull request merged**